File: go/kernel/memory_backend_test.go
// Copyright 2026 Oscar Yáñez Cisterna (@SkrOYC)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package kernel_test

import (
	"testing"

	kernel "github.com/tuvren/framework/go/kernel"
)

// TestInMemoryBackend_GetTurnTree_ReturnsDefensiveCopy is the P2-2
// regression: mutating a manifest returned by GetTurnTree (including an
// ordered PathValue's backing slice within it) must not corrupt the
// backend's stored state.
func TestInMemoryBackend_GetTurnTree_ReturnsDefensiveCopy(t *testing.T) {
	k := newTestKernel()
	if err := k.RegisterSchema(canonicalSchema()); err != nil {
		t.Fatalf("register: %v", err)
	}
	treeHash, err := k.CreateTurnTree("schema_main", map[string]kernel.PathValue{
		"messages":         {Kind: kernel.PathValueOrderedKind, Ordered: []string{"a", "b"}},
		"context.manifest": {Kind: kernel.PathValueNull},
	}, nil)
	if err != nil {
		t.Fatalf("create tree: %v", err)
	}

	first, ok := k.Backend.GetTurnTree(treeHash)
	if !ok {
		t.Fatalf("expected tree %q to exist", treeHash)
	}

	// Mutate the returned manifest map itself...
	first.Manifest["messages"] = kernel.PathValue{Kind: kernel.PathValueSingleKind, Single: "corrupted"}
	// ...and the backing slice of an ordered value obtained before that
	// overwrite.
	second, ok := k.Backend.GetTurnTree(treeHash)
	if !ok {
		t.Fatalf("expected tree %q to exist", treeHash)
	}
	second.Manifest["messages"].Ordered[0] = "corrupted-in-place"

	third, ok := k.Backend.GetTurnTree(treeHash)
	if !ok {
		t.Fatalf("expected tree %q to exist", treeHash)
	}
	if third.Manifest["messages"].Kind != kernel.PathValueOrderedKind {
		t.Fatalf("expected messages to remain an ordered value, got %+v", third.Manifest["messages"])
	}
	got := third.Manifest["messages"].Ordered
	if len(got) != 2 || got[0] != "a" || got[1] != "b" {
		t.Fatalf("expected the stored manifest to be unchanged by mutating returned copies, got %+v", got)
	}
}

// TestInMemoryBackend_GetTurnNode_ReturnsDefensiveCopy is the P2-2
// regression: mutating a ConsumedStagedResults slice returned by
// GetTurnNode must not corrupt the backend's stored state.
func TestInMemoryBackend_GetTurnNode_ReturnsDefensiveCopy(t *testing.T) {
	k := newTestKernel()
	if err := k.RegisterSchema(canonicalSchema()); err != nil {
		t.Fatalf("register: %v", err)
	}
	result, err := k.CreateThread("thread_defensive", "schema_main", "branch_main")
	if err != nil {
		t.Fatalf("create thread: %v", err)
	}
	steps := []kernel.StepDeclaration{{ID: "only_step", Deterministic: true, SideEffects: false}}
	if err := k.CreateRun("run_1", "turn_1", "branch_main", "schema_main", result.RootTurnNodeHash, steps); err != nil {
		t.Fatalf("create run: %v", err)
	}
	staged := kernel.StagedResult{
		TaskID: "task", ObjectHash: kernel.HashBytesToHex([]byte("message-defensive")),
		ObjectType: "message", Status: kernel.StagedResultCompleted, Timestamp: 1,
	}
	if err := k.StageResult("run_1", staged); err != nil {
		t.Fatalf("stage: %v", err)
	}
	nodeHash, err := k.CompleteStep("run_1", "only_step", "", "")
	if err != nil {
		t.Fatalf("complete step: %v", err)
	}

	first, ok := k.Backend.GetTurnNode(nodeHash)
	if !ok {
		t.Fatalf("expected turn node %q to exist", nodeHash)
	}
	if len(first.ConsumedStagedResults) != 1 {
		t.Fatalf("expected exactly one consumed staged result, got %d", len(first.ConsumedStagedResults))
	}

	// Mutate the returned slice's element in place, then append to it too.
	first.ConsumedStagedResults[0].TaskID = "corrupted"
	first.ConsumedStagedResults = append(first.ConsumedStagedResults, kernel.StagedResult{TaskID: "injected"})

	second, ok := k.Backend.GetTurnNode(nodeHash)
	if !ok {
		t.Fatalf("expected turn node %q to exist", nodeHash)
	}
	if len(second.ConsumedStagedResults) != 1 {
		t.Fatalf("expected the stored node to still have exactly one consumed staged result, got %d: %+v", len(second.ConsumedStagedResults), second.ConsumedStagedResults)
	}
	if second.ConsumedStagedResults[0].TaskID != "task" {
		t.Fatalf("expected the stored consumed staged result's taskId to be unchanged, got %q", second.ConsumedStagedResults[0].TaskID)
	}
}

// TestInMemoryBackend_ThreadRootOwnershipIndex proves PutThread /
// GetThreadByRootTurnNode's storage-level contract directly: the backend
// records which thread owns a given root turn node hash, independent of
// Kernel.CreateThread's own use of that index for the P0-1
// ErrThreadRootNotUnique guard.
func TestInMemoryBackend_ThreadRootOwnershipIndex(t *testing.T) {
	backend := kernel.NewInMemoryBackend(&kernel.FixedClock{Ms: 1})

	if _, ok := backend.GetThreadByRootTurnNode("unclaimed"); ok {
		t.Fatal("expected an unclaimed root hash to report no owner")
	}

	rootHash := kernel.HashBytesToHex([]byte("root-node"))
	if !backend.PutThread(kernel.Thread{ThreadID: "thread_owner", RootTurnNodeHash: rootHash, CreatedAtMs: 1}) {
		t.Fatal("expected PutThread to succeed for a fresh threadId")
	}
	owner, ok := backend.GetThreadByRootTurnNode(rootHash)
	if !ok || owner != "thread_owner" {
		t.Fatalf("expected thread_owner to own %q, got %q (ok=%v)", rootHash, owner, ok)
	}

	// Re-registering the same threadId is rejected regardless of root
	// ownership (PutThread's own "already exists" contract).
	if backend.PutThread(kernel.Thread{ThreadID: "thread_owner", RootTurnNodeHash: rootHash, CreatedAtMs: 2}) {
		t.Fatal("expected PutThread to reject a duplicate threadId")
	}
}

// TestAsKernelError_UnwrapsWrappedError is the P2-3 regression: AsKernelError
// must find a *KernelError through fmt.Errorf's %w wrapping (errors.As),
// not just a bare type assertion.
func TestAsKernelError_UnwrapsWrappedError(t *testing.T) {
	base := &kernel.KernelError{Code: "kernel_runtime_test_wrapped", Message: "wrapped"}
	wrapped := errorsJoinFmtWrap(base)

	kerr, ok := kernel.AsKernelError(wrapped)
	if !ok {
		t.Fatalf("expected AsKernelError to unwrap a wrapped *KernelError, got ok=false for %v", wrapped)
	}
	if kerr.Code != base.Code {
		t.Fatalf("expected code %q, got %q", base.Code, kerr.Code)
	}
}

// errorsJoinFmtWrap wraps err with fmt.Errorf's %w verb, isolated into a
// tiny helper so the test above reads as "wrap it" without importing fmt
// directly into the top-level test flow.
func errorsJoinFmtWrap(err error) error {
	return wrappedError{inner: err}
}

type wrappedError struct{ inner error }

func (w wrappedError) Error() string { return "wrapped: " + w.inner.Error() }
func (w wrappedError) Unwrap() error { return w.inner }

var _ error = wrappedError{}

File: dart/kernel/test/memory_backend_test.dart
// Copyright 2026 Oscar Yáñez Cisterna (@SkrOYC)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/// [InMemoryBackend] coverage, porting the essential scenarios from
/// `go/kernel/memory_backend_test.go` (defensive copies), the
/// `kernel.scope-isolation` section of
/// `go/kernel/scope_isolation_and_reclamation_test.go`, and
/// `go/kernel/thread_root_uniqueness_internal_test.go` (root ownership
/// index / genesis-hash-collision defense).
library;

import 'package:test/test.dart';
import 'package:tuvren_kernel/tuvren_kernel.dart';

import 'support/kernel_fixtures.dart';

(Kernel, Kernel) newScopedKernelPair() {
  final store = MemoryScopeStore();
  final clock = IncrementingClock();
  final backendA = InMemoryBackend.scoped(clock, store, 'tuvren.scope.test-a');
  final backendB = InMemoryBackend.scoped(clock, store, 'tuvren.scope.test-b');
  return (
    Kernel('scope-a', clock, backendA),
    Kernel('scope-b', clock, backendB),
  );
}

void main() {
  group('scope isolation', () {
    test('hasObject is scope-confined', () {
      final (kernelA, kernelB) = newScopedKernelPair();
      final hash = kernelA.putObject(
        'application/json',
        'scope-a content'.codeUnits,
      );
      expect(kernelA.hasObject(hash), isTrue);
      expect(kernelB.hasObject(hash), isFalse);
    });

    test('getObject is scope-confined', () {
      final (kernelA, kernelB) = newScopedKernelPair();
      final hash = kernelA.putObject(
        'application/json',
        'scope-a content'.codeUnits,
      );
      expect(kernelA.backend.getObject(hash), isNotNull);
      expect(kernelB.backend.getObject(hash), isNull);
    });

    test('thread enumeration is scope-confined', () {
      final (kernelA, kernelB) = newScopedKernelPair();
      kernelA.registerSchema(canonicalSchema());
      kernelA.createThread(
        'thread_scope_probe',
        'schema_main',
        'branch_scope_probe',
      );

      final (threadsA, _) = kernelA.listThreads(0, '');
      expect(threadsA.map((t) => t.threadId), contains('thread_scope_probe'));

      final (threadsB, _) = kernelB.listThreads(0, '');
      expect(
        threadsB.map((t) => t.threadId),
        isNot(contains('thread_scope_probe')),
      );
    });

    test(
      'two handles bound to the same store and scope share committed state',
      () {
        final store = MemoryScopeStore();
        final clock = IncrementingClock();
        final backend1 = InMemoryBackend.scoped(
          clock,
          store,
          'tuvren.scope.shared',
        );
        final backend2 = InMemoryBackend.scoped(
          clock,
          store,
          'tuvren.scope.shared',
        );
        final kernel1 = Kernel('shared', clock, backend1);
        final kernel2 = Kernel('shared', clock, backend2);

        final hash = kernel1.putObject(
          'application/json',
          'shared content'.codeUnits,
        );
        expect(kernel2.hasObject(hash), isTrue);
      },
    );
  });

  group('defensive copies', () {
    test('getTurnTree returns a copy independent of later mutation', () {
      final kernel = newTestKernel();
      kernel.registerSchema(canonicalSchema());
      final hash = kernel.createTurnTree('schema_main', {
        'messages': const PathValue.ordered([]),
        'context.manifest': const PathValue.nullValue(),
      });

      final first = kernel.backend.getTurnTree(hash)!;
      first.manifest['messages'] = PathValue.ordered(['injected']);

      final second = kernel.backend.getTurnTree(hash)!;
      expect(second.manifest['messages']!.ordered, isEmpty);
    });

    test('mutating a returned ordered PathValue list in place does not '
        'corrupt stored state', () {
      // Regression for the P2-2 bug: TurnTree.clone() used to do a shallow
      // Map.of(manifest), so an ordered PathValue's backing List<String>
      // was shared between the backend's stored state and every value
      // returned by getTurnTree -- mutating the returned list in place
      // (rather than replacing the map entry) corrupted the stored tree.
      final kernel = newTestKernel();
      kernel.registerSchema(canonicalSchema());
      final hash = kernel.createTurnTree('schema_main', {
        'messages': const PathValue.ordered(['a', 'b']),
        'context.manifest': const PathValue.nullValue(),
      });

      final first = kernel.backend.getTurnTree(hash)!;
      first.manifest['messages']!.ordered!.add('injected');

      final second = kernel.backend.getTurnTree(hash)!;
      expect(second.manifest['messages']!.ordered, equals(['a', 'b']));
    });

    test(
      'getTurnNode returns a copy whose consumedStagedResults is independent storage',
      () {
        final kernel = newTestKernel();
        createSingleStepRun(kernel, 'thread_a', 'branch_a', 'run_a');
        final hash = kernel.completeStep('run_a', 'only_step', '', '');

        final first = kernel.backend.getTurnNode(hash)!;
        first.consumedStagedResults.add(
          const StagedResult(
            taskId: 'injected',
            objectHash:
                '0000000000000000000000000000000000000000000000000000000000000000',
            objectType: 'message',
            timestamp: 0,
            status: StagedResultStatus.completed,
          ),
        );

        final second = kernel.backend.getTurnNode(hash)!;
        expect(second.consumedStagedResults, isEmpty);
      },
    );
  });

  group('thread root uniqueness', () {
    // go/kernel/thread_root_uniqueness_internal_test.go exercises
    // createThread's ErrThreadRootNotUnique defense-in-depth guard as an
    // *internal* (same-package) test by directly manipulating unexported
    // backend state to force a genesis-hash collision -- something
    // structurally unreachable through normal use once genesis nodes carry
    // a thread-unique bootstrap eventHash (see createThread's doc
    // comment). This port has no test-only backdoor into InMemoryBackend's
    // private scope state to reproduce that exact white-box setup, so this
    // instead pins the root-ownership index the guard depends on: every
    // created thread's root hash resolves back to its own owning thread,
    // and two threads on the same schema never collide.
    test('the root-ownership index resolves each thread back to itself', () {
      final kernel = newTestKernel();
      kernel.registerSchema(canonicalSchema());
      final a = kernel.createThread('thread_a', 'schema_main', 'branch_a');
      final b = kernel.createThread('thread_b', 'schema_main', 'branch_b');

      expect(
        kernel.backend.getThreadByRootTurnNode(a.rootTurnNodeHash),
        'thread_a',
      );
      expect(
        kernel.backend.getThreadByRootTurnNode(b.rootTurnNodeHash),
        'thread_b',
      );
      expect(a.rootTurnNodeHash, isNot(equals(b.rootTurnNodeHash)));
    });
  });
}
